### PR TITLE
OV encoding

### PR DESCRIPTION
**Pull request by mojiiba**
This pull addressed https://github.com/openconfig/bootz/issues/40. Note that the pull is based on pull https://github.com/openconfig/bootz/pull/37.

### CODE REVIEW

File: server/entitymanager/entitymanager.go [hunk 1]
```diff
@@ -57,14 +62,40 @@ type InMemoryEntityManager struct {
 }
 
 // ResolveChassis returns an entity based on the provided lookup.
-func (m *InMemoryEntityManager) ResolveChassis(chassis *service.EntityLookup) (*service.ChassisEntity, error) {
+// In cases when the serail for  moulder chassis is not set, it uses the controller card to find the chasiss.
+func (m *InMemoryEntityManager) ResolveChassis(lookup *service.EntityLookup, ccSerail string) (*service.ChassisEntity, error) {
```
Comment (gmacf) [inline server/entitymanager/entitymanager.go:66]: Type in serial

File: server/entitymanager/entitymanager.go [hunk 2]
```diff
@@ -57,14 +62,40 @@ type InMemoryEntityManager struct {
...
+			if c.GetSerialNumber() == ccSerial {
+				if ch.Manufacturer != lookup.Manufacturer {
+					continue
+				}
+				return ch, nil
+			}
+		}
 	}
-	return &service.ChassisEntity{BootMode: ch.GetBootMode()}, nil
+	return nil, status.Errorf(codes.NotFound, "could not find chasis for controller card with serial# %s:", ccSerial)
```
Comment (gmacf) [inline server/entitymanager/entitymanager.go:98]: Typo in chassis

File: server/entitymanager/entitymanager.go [hunk 3]
```diff
@@ -102,43 +133,63 @@ func populateBootConfig(conf *epb.BootConfig) (*bootz.BootConfig, error) {
 }
 
 // GetBootstrapData fetches and returns the bootstrap data response from the server.
-func (m *InMemoryEntityManager) GetBootstrapData(chassis *service.EntityLookup, controllerCard *bootz.ControlCard) (*bootz.BootstrapDataResponse, error) {
+func (m *InMemoryEntityManager) GetBootstrapData(lo *service.EntityLookup, controllerCard *bootz.ControlCard) (*bootz.BootstrapDataResponse, error) {
```
Comment (gmacf) [inline server/entitymanager/entitymanager.go:136]: Nit: I would pick a better parameter name than 'lo'. Maybe el or l.

File: server/entitymanager/entitymanager.go [hunk 4]
```diff
@@ -102,43 +133,63 @@ func populateBootConfig(conf *epb.BootConfig) (*bootz.BootConfig, error) {
...
 	// First check if we are expecting this control card.
-	if controllerCard.SerialNumber == "" {
-		return nil, status.Errorf(codes.InvalidArgument, "no serial number provided")
+	serial := ""
+	fixedChasis := false
+	if controllerCard == nil {
+		if lo.SerialNumber == "" {
+			return nil, status.Errorf(codes.NotFound, "chassis type (fixed/modular) can not be determined, either controller card or chassis serial must be set ")
```
Comment (gmacf) [inline server/entitymanager/entitymanager.go:142]: Probably codes.invalidargument is more applicable here

File: server/entitymanager/entitymanager.go [hunk 5]
```diff
@@ -219,6 +270,11 @@ func parseSecurityArtifacts(artifactDir string) (*service.SecurityArtifacts, err
 	}, nil
 }
 
+// ssBase64 check if a string is base64 encoded.
+func isBase64(str string) bool {
```
Comment (gmacf) [inline server/entitymanager/entitymanager.go:274]: I might be wrong but I thought we agreed that we wouldn't base64 encode the OV if it's passed as bytes. Therefore doesn't that make this check unneeded?

**Comment by gmacf** [comment]
Should I also be reviewing PR 37? It looks like this one already is based off that.

File: server/entitymanager/entitymanager.go (hunk 5 above)
Comment (mojiiba) [reply server/entitymanager/entitymanager.go:274]: yes, here I am checking if  user provided ov is base64 (in inventory file), then we decode and send it as binary to the router.

File: server/entitymanager/entitymanager.go (hunk 4 above)
Comment (mojiiba) [reply server/entitymanager/entitymanager.go:142]: sure

File: server/entitymanager/entitymanager.go (hunk 3 above)
Comment (mojiiba) [reply server/entitymanager/entitymanager.go:136]: sure

**Comment by mojiiba** [comment]
> Should I also be reviewing PR 37? It looks like this one already is based off that.

no, as u said this one is based on that pr. I created two pr to ensure one pr per issue.

**Comment by mojiiba** [comment]
@mojiiba addressed your comments, thanks for the review.

File: server/entitymanager/entitymanager.go [hunk 6]
```diff
@@ -57,14 +62,40 @@ type InMemoryEntityManager struct {
...
+			chassis = ch
+		} else {
+			return nil, status.Errorf(codes.NotFound, "Could not find chassis with serial#: %s and manufacturer: %s and controller card %s",
+				lookup.SerialNumber, lookup.Manufacturer, ccSerial)
+		}
+	}
+	return &service.ChassisEntity{BootMode: chassis.GetBootMode()}, nil
+}
+
+// find controller card
```
Comment (melzhan) [inline server/entitymanager/entitymanager.go:86]: Comments should be full sentences and should always be capitalized and punctuated.
i.e. // Find controller card.

File: server/entitymanager/entitymanager.go [hunk 7]
```diff
@@ -275,18 +338,26 @@ func (m *InMemoryEntityManager) Sign(resp *bpb.GetBootstrapDataResponse, chassis
...
 	}
-	//cc:=&bpb.ControlCard{}
-	for _, c := range ch.GetControllerCards() {
+	for _, c := range chassis.GetControllerCards() {
 		if c.GetSerialNumber() == ccSerial {
 			return c.GetOwnershipVoucher(), nil
 		}
 	}
-	return "", status.Errorf(codes.NotFound, "could not find Controller with serial#: %s belonging to chassis %s", ccSerial, chassis.SerialNumber)
+	// hanlde fixed chassis
```
Comment (melzhan) [inline server/entitymanager/entitymanager.go:356]: Typo, please fix to:
Handle fixed chassis.

File: server/entitymanager/entitymanager_test.go [hunk 8]
```diff
@@ -403,22 +409,49 @@ func TestGetBootstrapData(t *testing.T) {
...
+		desc                string
+		input               *bpb.ControlCard
+		chassisSerial       string
+		chassisManufacturer string
+		want                *bpb.BootstrapDataResponse
+		wantErr             bool
 	}{{
-		desc:    "No serial number",
-		input:   &bpb.ControlCard{},
+		desc:                "No controller card, but valid chasis (success)",
```
Comment (melzhan) [inline server/entitymanager/entitymanager_test.go:430]: Typo: chassis

File: server/entitymanager/entitymanager_test.go [hunk 9]
```diff
@@ -403,22 +409,49 @@ func TestGetBootstrapData(t *testing.T) {
...
+			ServerTrustCert:  "FakeTLSCert",
+			BootConfig: &bpb.BootConfig{
+				VendorConfig: []byte(""),
+				OcConfig:     []byte(""),
+			},
+			Credentials: &bpb.Credentials{},
+		},
+		wantErr: false,
+	}, {
+		desc:    "No controller card and no chaisis serail (fail)",
```
Comment (melzhan) [inline server/entitymanager/entitymanager_test.go:453]: Typo: chassis serial

File: server/entitymanager/entitymanager_test.go [hunk 10]
```diff
@@ -427,11 +460,39 @@ func TestGetBootstrapData(t *testing.T) {
 		},
 		wantErr: true,
 	}, {
-		desc: "Successful bootstrap",
+		desc: "Successful bootstrap, valid chasis serial and controller card",
```
Comment (melzhan) [inline server/entitymanager/entitymanager_test.go:463]: Typo: chassis

File: server/entitymanager/entitymanager_test.go [hunk 11]
```diff
@@ -427,11 +460,39 @@ func TestGetBootstrapData(t *testing.T) {
...
+			ServerTrustCert:  "FakeTLSCert",
+			BootConfig: &bpb.BootConfig{
+				VendorConfig: []byte(""),
+				OcConfig:     []byte(""),
+			},
+			Credentials: &bpb.Credentials{},
+		},
+		wantErr: false,
+	}, {
+		desc: "Successful bootstrap, no chasis serial but valid controller card",
```
Comment (melzhan) [inline server/entitymanager/entitymanager_test.go:489]: typo: chassis

File: server/entitymanager/entitymanager_test.go [hunk 12]
```diff
@@ -450,6 +511,15 @@ func TestGetBootstrapData(t *testing.T) {
 			Credentials: &bpb.Credentials{},
 		},
 		wantErr: false,
+	}, {
+		desc: "Unsuccessful bootstrap, no chasis serial, valid controller card, not matching manufacturer",
```
Comment (melzhan) [inline server/entitymanager/entitymanager_test.go:515]: typo: chassis

File: proto/bootz.proto [hunk 13]
```diff
@@ -61,8 +61,8 @@ message GetBootstrapDataRequest {
 message ChassisDescriptor {
   string manufacturer = 1;
   string part_number = 2;
-  // This field is populated for fixed form-factor chassis, and empty for
-  // modular chassis.
+  // This field must be populated for fixed form-factor chassis.
+  // the serial can be empty for modular chassis.
```
Comment (melzhan) [inline proto/bootz.proto:65]: Capitalize the sentence

File: client/client.go [hunk 14]
```diff
@@ -87,16 +87,9 @@ func validateArtifacts(serialNumber string, resp *bpb.GetBootstrapDataResponse,
 		return fmt.Errorf("received empty ownership certificate from server")
 	}
```
Comment (melzhan) [inline client/client.go:89]: Hey Moji, just letting you know we are working on importing this code into Google internally, which has strict code quality standards. I'll be leaving comments based on what needs to be updated such that the code can pass our internal linters.

Comment (mojiiba) [reply client/client.go:89]: @melzhan thanks for the detailed review, addressed the comments.

File: server/entitymanager/entitymanager.go (hunk 5 above)
Comment (gmacf) [reply server/entitymanager/entitymanager.go:274]: I see, thanks.

File: server/entitymanager/entitymanager.go [hunk 15]
```diff
@@ -57,14 +62,40 @@ type InMemoryEntityManager struct {
 }
 
 // ResolveChassis returns an entity based on the provided lookup.
-func (m *InMemoryEntityManager) ResolveChassis(chassis *service.EntityLookup) (*service.ChassisEntity, error) {
+// In cases when the serial for moduldar chassis is not set, it uses the controller card to find the chassis.
```
Comment (gmacf) [inline server/entitymanager/entitymanager.go:65]: Typo in "modular"

File: server/entitymanager/entitymanager.go [hunk 16]
```diff
@@ -353,7 +353,7 @@ func (m *InMemoryEntityManager) fetchOwnershipVoucher(lookup *service.EntityLook
 			return c.GetOwnershipVoucher(), nil
 		}
 	}
-	// hanlde fixed chassis
+	//Handle fixed chassis.
```
Comment (melzhan) [inline server/entitymanager/entitymanager.go:356]: missing a space between // and Handle

**Comment by gmacf** [comment]
Still a few typos in the code (both variable names and comments). Please fix those up but apart from that the logic looks good. Please also resolve the comments when they're addressed.

File: server/entitymanager/entitymanager.go [hunk 17]
```diff
@@ -83,7 +83,7 @@ func (m *InMemoryEntityManager) ResolveChassis(lookup *service.EntityLookup, ccS
 	return &service.ChassisEntity{BootMode: chassis.GetBootMode()}, nil
 }
 
-// find controller card
+// resolveChassisViaControllerCard resolves a chassis based on controller card serial
```
Comment (melzhan) [inline server/entitymanager/entitymanager.go:86]: Missing full stop

File: server/entitymanager/entitymanager.go [hunk 18]
```diff
@@ -219,6 +270,11 @@ func parseSecurityArtifacts(artifactDir string) (*service.SecurityArtifacts, err
 	}, nil
 }
 
+// ssBase64 check if a string is base64 encoded.
```
Comment (melzhan) [inline server/entitymanager/entitymanager.go:273]: typo: isBase64

File: server/entitymanager/entitymanager.go [hunk 19]
```diff
@@ -275,18 +338,26 @@ func (m *InMemoryEntityManager) Sign(resp *bpb.GetBootstrapDataResponse, chassis
 }
 
 // FetchOwnershipVoucher retrieves the ownership voucher for a control card
```
Comment (melzhan) [inline server/entitymanager/entitymanager.go:340]: typo: fetchOwnershipVoucher

**Review by gmacf — APPROVED** [review]

**Pull request merged**